cloud/services/virtualmachineextensions/virtualmachineextensions_test.go

```diff
@@ -0,0 +1,398 @@
+					t.Fatalf("got an unexpected error: %v", err)
+				}
+			} else {
+				if tc.expectedError != "" {
+					t.Fatalf("expected an error: %v", tc.expectedError)
+
+				}
+			}
+		})
+	}
+}
+
+func TestReconcileVMExtensions(t *testing.T) {
+	testcases := []struct {
+		name            string
+		vmExtensionSpec Spec
+		expectedError   string
+		expect          func(m *mock_virtualmachineextensions.MockClientMockRecorder)
+	}{
+		{
+			name: "vm extension create successfully",
+			vmExtensionSpec: Spec{
+				Name:       "my-vmext",
+				VMName:     "my-vm",
+				ScriptData: "",
+			},
+			expectedError: "",
+			expect: func(m *mock_virtualmachineextensions.MockClientMockRecorder) {
+				m.CreateOrUpdate(context.TODO(), "my-rg", "my-vm", "my-vmext", gomock.AssignableToTypeOf(compute.VirtualMachineExtension{}))
+			},
+		},
+		{
+			name: "fail to create a vm extension",
+			vmExtensionSpec: Spec{
+				Name:       "my-vmext",
+				VMName:     "my-vm",
+				ScriptData: "",
+			},
+			expectedError: "cannot create vm extension: #: Internal Server Error: StatusCode=500",
+			expect: func(m *mock_virtualmachineextensions.MockClientMockRecorder) {
+				m.CreateOrUpdate(context.TODO(), "my-rg", "my-vm", "my-vmext", gomock.AssignableToTypeOf(compute.VirtualMachineExtension{})).Return(autorest.NewErrorWithResponse("", "", &http.Response{StatusCode: 500}, "Internal Server Error"))
+			},
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockCtrl := gomock.NewController(t)
+			vmExtensionMock := mock_virtualmachineextensions.NewMockClient(mockCtrl)
+
+			cluster := &clusterv1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{Name: "test-cluster"},
+			}
+
+			client := fake.NewFakeClient(cluster)
+
+			tc.expect(vmExtensionMock.EXPECT())
+
+			clusterScope, err := scope.NewClusterScope(scope.ClusterScopeParams{
+				AzureClients: scope.AzureClients{
+					SubscriptionID: "123",
+					Authorizer:     autorest.NullAuthorizer{},
+				},
+				Client:  client,
+				Cluster: cluster,
+				AzureCluster: &infrav1.AzureCluster{
+					Spec: infrav1.AzureClusterSpec{
+						Location: "test-location",
+						ResourceGroup: "my-rg",
+						NetworkSpec: infrav1.NetworkSpec{
+							Vnet: infrav1.VnetSpec{
+								ID:            "my-vnet-id",
+								Name:          "my-vnet",
+								ResourceGroup: "my-rg",
+							},
+						},
+					},
+				},
+			})
+			if err != nil {
+				t.Fatalf("Failed to create test context: %v", err)
+			}
+
+			s := &Service{
+				Scope:  clusterScope,
+				Client: vmExtensionMock,
+			}
+
+			if err := s.Reconcile(context.TODO(), &tc.vmExtensionSpec); err != nil {
+				if tc.expectedError == "" || err.Error() != tc.expectedError {
+					t.Fatalf("got an unexpected error: %v", err)
+				}
+			} else {
+				if tc.expectedError != "" {
+					t.Fatalf("expected an error: %v", tc.expectedError)
+
+				}
+			}
+		})
+	}
+}
+
+func TestDeleteRouteTable(t *testing.T) {
```

Comment:
```suggestion
func TestDeleteVMExtension(t *testing.T) {
```

Reply:
sorry :( too much copy +paste

Reply:
done

Reply:
thanks for the review and catch @CecileRobertMichon